### ISSUE TITLE
Build core.py: the Protein class, io/, and ADR-0001

### ISSUE THREAD

**Issue by lhqing** [liuhlab/liulab-protein#10]
## Question

Build `core.py`, the `io/` package, and ADR-0001.

Spec: the `core.py` section of #1, plus the issue comment asking that io get its own
directory — FASTA now, mmCIF and other file types later.

Deliver on the `v1` branch:

- `src/protein/core.py` — **`class Protein(EmbeddingMixin, SearchMixin)`**, two mixins, not
  three. `StructureMixin` dissolved in #9: foldseek takes a structure and a `Protein` has no
  coordinates, so `.structure`, `foldseek_search()` and `from_structure(path)` are **not** on
  this class. That is the map's *direct support only* rule.
- Follows `Genome`'s shape exactly: mixins first in the bases, each using
  `cast("Protein", self)` under a `TYPE_CHECKING` import and deferring heavy imports into
  method bodies. **Importing `protein` must never import torch** — assert it in a test.
- Surface: `.sequence`, `.id`, `.name`, `.description`, `.metadata`, `.length`, `__len__`,
  `__getitem__` returning `str` (a slice of a protein is not a protein), `__repr__` →
  `Protein('P12345', 214 aa)`, `from_fasta`, `to_fasta`. Metadata eager and total; everything
  expensive lazy and cached on the instance.
- It validates at construction.
- `src/protein/io/` — `__init__.py` and `fasta.py`. The shape must leave room for
  `structure.py` without moving anything.
- Stub **two** mixin modules so the bases resolve; the lanes fill them in.
- `docs/adr/0001-*.md` — validate at construction. This departs from `liulab-genome`'s
  ADR-0005, which declines validation because scanning a chromosome is expensive; a protein
  is hundreds of residues, so that cost argument does not carry. Hard to reverse, surprising
  given the sibling, a real trade-off — all three, so it earns an ADR. 400-word cap, measured
  with `pixi run vale`, never `wc -w`. Agent-facing: needs `search: exclude: true` **and**
  absence from `nav:`, per conformance rule 2.

**Blocked on #22.** Whether `.sequence` is a `str` or a biotite `ProteinSequence` is being
decided there, and it changes this class's constructor and `__getitem__`. Do not start until
it resolves.

Read `~/src/liulab-genome/src/genome/assembly/genome.py` before writing anything.

**Comment by lhqing** [comment]
Changed by [What is a structure in this package?](https://github.com/liuhlab/liulab-protein/issues/9).

The class declaration in #1 is now:

```python
class Protein(EmbeddingMixin, SearchMixin):
```

`StructureMixin` is gone. `Protein` keeps `embed()` and `search()` — ESM-C and mmseqs take a sequence directly — and loses `.structure` and `foldseek_search()`, because foldseek takes a structure and a `Protein` has none. That is the map's **direct support only** rule, now in the Notes.

**`Protein.from_structure(path)` is also dropped**, replaced by `Structure.from_file(path)` in the structure lane. A `Protein` is identified by a UniProt accession, which a structure file does not carry.

Everything else in this ticket stands: the signature, the `cast("Protein", self)` pattern for the two remaining mixins, the never-import-torch rule, and ADR-0001 on validating at construction. Stub only the two mixins now, not three.

**Comment by lhqing** [comment]
Body rewritten 2026-09-02 to absorb the amendment comment above — two mixins, not three, and no `.structure`/`foldseek_search`/`from_structure` on `Protein`. Also flags that #22 decides whether `.sequence` is a `str` or a biotite `ProteinSequence`.

**Comment by lhqing** [comment]
**Unblocked, and it gains a second ADR.** From #22:

- **`.sequence` is a biotite `ProteinSequence`, not a `str`.** `str(p.sequence)` is how ESM-C
  and mmseqs get their string; `p.sequence == "MKT"` is `False`. The divergence from
  `liulab-genome`'s `str`-subclass `DNA` is deliberate and accepted — genome slices chromosomes
  and needs `str` cheapness, protein carries a few hundred residues into typed biotite calls.
  `__getitem__` still returns `str`: a slice of a protein is not a protein.
- **Write ADR-0002: biotite's file and array layer only, never its convenience converters.**
  `fasta.get_sequence` and `structure.to_sequence` map `U`→`C` and `O`→`K` silently, and this
  package coerces to `X` at its own boundary instead. It earns an ADR on all three counts:
  surprising (the obvious helper is the wrong one), silent when violated, and a real trade-off
  (we drop `U`/`O` fidelity deliberately). 400-word cap, `search: exclude: true`, absent from
  `nav:`.
- `io/fasta.py` adopts biotite's `FastaFile` / `read_iter` — the UniProt header survives
  byte-for-byte, and 575,503 entries stream in 0.9 s at 0.05 GB. **gzip is ours**: biotite has
  no gzip path and no extension sniffing, but a `gzip.open(p, "rt")` file object works. Three
  lines.
- The string→sequence step is ours, not `fasta.get_sequence` — that one guesses
  `NucleotideSequence` first, so `"MKTU"` comes back as *DNA*.

**Comment by lhqing** [comment]
## Landed on branch `i10` (`6a218d3`), on top of #11's `5c2ac4e`

`src/protein/core.py`, `src/protein/io/{__init__,fasta}.py`,
`src/protein/search/{__init__,mixin}.py`, `docs/adr/0001-*.md`, `docs/adr/0002-*.md`,
`tests/test_core.py`, `tests/test_io_fasta.py`, two FASTA fixtures with provenance in
`tests/data/README.md`. Four peripheral edits, listed at the bottom.

### This ticket's body disagreed with #1, and #1 won

#10 says `class Protein(EmbeddingMixin, SearchMixin)` and "two mixins, not three". That
predates #7. **#7 dissolved `EmbeddingMixin`** and #1 was amended to `class
Protein(SearchMixin)`. I followed #1: **one mixin, one stub module**, and `Protein` has no
`embed()`. #17 should amend this body when it sweeps the spec.

### Surface

```python
Protein(sequence, *, id=None, name=None, description=None, metadata=None)
.sequence  .id  .name  .description  .metadata  .length
__len__  __getitem__ -> str  __repr__ -> "Protein('P12345', 214 aa)"
from_fasta(path) -> Self          to_fasta(path=None, *, line_width=80) -> str
```

`.sequence` is a biotite `ProteinSequence` (#22), built by
`seq.to_protein_sequence(text, name=id)` and never by `ProteinSequence(text)` — #8's closing
comment is what the constructor obeys, and it is what makes the guard unreachable-around.
`.metadata` is eager, total and a copy. Nothing here is expensive, so nothing is lazily
cached: the class holds one already-parsed sequence and four labels. No `.structure`, no
`foldseek_search()`, no `from_structure()`, no `embed()` — asserted by a parametrised test.

`to_fasta` returns the text and *also* writes when given a path, so a doctest can show it and
a caller can skip the `StringIO`. `from_fasta` reads **exactly one** record and refuses a
multi-record file by name, pointing at `io.fasta.read_proteins` — taking the first silently is
the failure nobody notices.

### The `SearchMixin` stub #14 has to fit

`src/protein/search/mixin.py` holds **`class SearchMixin:` with a docstring and no method**,
re-exported from `src/protein/search/__init__.py` as `__all__ = ["SearchMixin"]`. Empty rather
than a `search` that raises, so `hasattr(p, "search")` does not lie. #14 adds

```python
def search(self, database: SequenceDatabase | str, **kwargs: Any) -> pd.DataFrame
```

there, with `cast("Protein", self)` under `TYPE_CHECKING` and pandas imported inside the
method body. Nothing else in the lane exists yet, so `mmseqs.py`, `foldseek.py` and `cli.py`
are #14's to add, and #11's commented `add_typer` block in `src/protein/cli.py` uncomments as
written — `from protein.search import cli as _search_cli` needs no change here.

### `io/` is a directory, and biotite's converters are refused by a test

`protein.io` exports the **submodule**, not its functions (`from protein.io import fasta`), so
`structure.py` lands beside `fasta.py` for #15 and nothing moves — and two formats both
wanting `read_records` never collide.

`fasta.py` is two layers. The record layer is biotite's `FastaFile.read_iter` /
`write_iter` — measured: a header survives byte-for-byte, runs of spaces included — plus the
gzip branch biotite has no reader for. The protein layer is ours:
`read_records`, `write_records`, `format_records`, `read_proteins`, `write_proteins`,
`split_header`, `join_header`.

`FastaFile.read_iter` and `write_iter` are **methods on `FastaFile`**, not module-level
functions, in biotite 1.4.0 — #1's table reads as if they were module functions.

**ADR-0002 is enforced, not just written.** `tests/test_io_fasta.py` parses every
`src/protein/**/*.py` with `ast` and fails on any `get_sequence`, `get_sequences` or
`to_sequence` reference in code. Parsed rather than grepped, so this repo's own prose about
the rule is not an offence, and a test proves the guard fires.

### One thing #12 inherits

`split_header` does the **plain FASTA** split only: first whitespace token is `id`, the rest
is `description`. So a Swiss-Prot header lands as `id == "sp|P01308|INS_HUMAN"`, and
`Protein.name` is never filled from a file. Resolving `db|accession|entry_name` is a
Swiss-Prot convention, not a FASTA one, and it belongs with the header-to-`metadata` mapping
the map still lists under *Not yet specified*. `split_header` is public so `SwissProt` can
build on it.

### Both ADRs, and why 0002 is here

`docs/adr/` did not exist; this establishes it. ADR-0001 argues with what genome's ADR-0005
actually says and answers both its reasons: biotite walks the string to build a
`ProteinSequence` anyway, so the scan is free, and the code a protein database writes for
uncertainty — `X` — is in this alphabet, where genome's `N` runs are not in DNA's. ADR-0002 is
#1's docs table assigning it to this ticket and the map calling it #10's job; it is the rule
`io/fasta.py` obeys, so writing it elsewhere would have meant shipping the code before the
record. Both carry `search: exclude: true`, neither is in `nav:`, and both pass
`Lab.LengthAdr` under `pixi run vale`.

### Edits outside my file set, for the merge

- `src/protein/__init__.py` — `Protein` added to the import and to `__all__`; nothing else.
- `docs/api.md`, `README.md`, `docs/index.md` — the dead `greet` references. The
  `docs/api.md` one is the sharp one: it is an mkdocstrings block, `docs-build` is a separate
  CI job, so it would have failed the pull request while `pixi run check` stayed green.
- `src/protein/cli.py` and `tests/test_placeholder.py` needed nothing: #11 had already
  dropped the `greet` verb and the placeholder test, so removing `greet` from `core.py` broke
  no code. `grep -r greet` over the tracked tree now returns nothing.
- **`CONTEXT.md` is still an empty glossary.** `Protein` belongs in it and I left it alone
  rather than collide with #15 and #16 in a small file.

### Gate, on GPU71FM in `~/pkg/protein-wt/i10` at `origin/i10`

```text
  pass  lint (0s)
  pass  fmt-check (0s)
  pass  typecheck (2s)
  pass  vale (11s)
  pass  markdownlint (1s)
  pass  conformance (0s)

All 6 steps passed in 11s.
...
collected 294 items

tests/test_cli.py ........                                               [  2%]
tests/test_conformance.py .............................................. [ 18%]
.........................................................                [ 37%]
tests/test_core.py ...............................                       [ 48%]
tests/test_external.py ................................................. [ 64%]
.                                                                        [ 65%]
tests/test_guards.py ....                                                [ 66%]
tests/test_io_fasta.py ..................                                [ 72%]
tests/test_seq.py ........................                               [ 80%]
tests/test_store.py .........                                            [ 84%]
src/protein/cli.py .                                                     [ 84%]
src/protein/core.py ....s..                                              [ 86%]
src/protein/external.py ........s.........s....s.                        [ 95%]
src/protein/io/fasta.py ...ss.                                           [ 97%]
src/protein/seq.py .....                                                 [ 98%]
src/protein/store.py ...                                                 [100%]

================= 288 passed, 6 skipped, 2 warnings in 16.47s ==================
```

`pixi run docs-build` is green too — `zensical build --clean --strict`, "No issues found".

The torch rule is checked where it can actually fail: in the `esm` environment, which has
torch 2.11 installed, `import protein` leaves `sys.modules` without it.

**Comment by lhqing** [comment]
Built and pushed on branch `i10` (`6a218d3`); the gate and `docs-build` are both green on GPU71FM. The orchestrator merges the branch.
